File: android/app/src/main/java/net/wastu/binderclip/DiagnosticLog.kt
package net.wastu.binderclip

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow

/** Redacted, persistent diagnostic trail. Clipboard bytes, filenames, keys and invitations never enter it. */
data class DiagnosticEvent(
    val timestamp: Long = System.currentTimeMillis(),
    val level: DiagnosticLevel,
    val message: String,
)

enum class DiagnosticLevel { Info, Warning, Error }

object DiagnosticLog {
    private const val MAXIMUM_EVENTS = 2_000
    private const val RETENTION_MS = 7 * 24 * 60 * 60 * 1_000L
    private const val PREFERENCES = "binderclip_diagnostics"
    private const val EVENTS = "events"
    private val lock = Any()
    private val mutableEvents = MutableStateFlow<List<DiagnosticEvent>>(emptyList())
    val events = mutableEvents.asStateFlow()
    @Volatile private var context: Context? = null

    fun initialize(context: Context) {
        synchronized(lock) {
        if (this.context != null) return
        this.context = context.applicationContext
        val now = System.currentTimeMillis()
        val restored = runCatching {
            JSONArray(context.getSharedPreferences(PREFERENCES, Context.MODE_PRIVATE).getString(EVENTS, "[]"))
        }.getOrElse { JSONArray() }
        mutableEvents.value = buildList {
            for (index in 0 until restored.length()) {
                val value = restored.optJSONObject(index) ?: continue
                val timestamp = value.optLong("timestamp")
                val level = value.optString("level").let { runCatching { DiagnosticLevel.valueOf(it) }.getOrNull() } ?: continue
                val message = value.optString("message").takeIf { it.isNotBlank() } ?: continue
                if (timestamp >= now - RETENTION_MS) add(DiagnosticEvent(timestamp, level, message))
            }
        }.takeLast(MAXIMUM_EVENTS)
        }
    }

    fun info(message: String) = append(DiagnosticLevel.Info, message)
    fun warning(message: String) = append(DiagnosticLevel.Warning, message)
    fun error(message: String) = append(DiagnosticLevel.Error, message)

    fun clear() { synchronized(lock) { mutableEvents.value = emptyList(); persistLocked() } }

    private fun append(level: DiagnosticLevel, message: String) {
        synchronized(lock) {
            val event = DiagnosticEvent(level = level, message = message)
            val previous = mutableEvents.value.lastOrNull()
            if (previous?.level == level && previous.message == message && event.timestamp - previous.timestamp < 10_000) return
            mutableEvents.value = (mutableEvents.value + event).filter { it.timestamp >= event.timestamp - RETENTION_MS }.takeLast(MAXIMUM_EVENTS)
            persistLocked()
        }
    }

    private fun persistLocked() {
        val appContext = context ?: return
        val serialized = JSONArray().also { array -> mutableEvents.value.forEach { event ->
            array.put(JSONObject().put("timestamp", event.timestamp).put("level", event.level.name).put("message", event.message))
        } }
        appContext.getSharedPreferences(PREFERENCES, Context.MODE_PRIVATE).edit().putString(EVENTS, serialized.toString()).apply()
    }
}

File: macos/BinderClipMac/Sources/Direct/DiagnosticLog.swift
import Foundation
import os

enum DiagnosticLevel: String, Codable {
    case info = "Info"
    case warning = "Warning"
    case error = "Error"
}

struct DiagnosticEvent: Codable {
    let date: Date
    let level: DiagnosticLevel
    let message: String
}

/// Redacted persistent log: it never contains clipboard data, filenames, keys, or invitations.
final class DiagnosticLog {
    static let shared = DiagnosticLog()
    static let changed = Notification.Name("net.wastu.binderclip.diagnosticsChanged")

    private let lock = NSLock()
    private let logger = Logger(subsystem: "net.wastu.binderclip", category: "Diagnostics")
    private var events: [DiagnosticEvent] = []
    private let storageKey = "diagnostic-events-v1"
    private let maximumEvents = 2_000
    private let retention: TimeInterval = 7 * 24 * 60 * 60

    private init() {
        events = (try? JSONDecoder().decode([DiagnosticEvent].self, from: UserDefaults.standard.data(forKey: storageKey) ?? Data()))?
            .filter { $0.date >= Date().addingTimeInterval(-retention) }.suffix(maximumEvents).map { $0 } ?? []
    }

    func info(_ message: String) { append(.info, message) }
    func warning(_ message: String) { append(.warning, message) }
    func error(_ message: String) { append(.error, message) }

    func snapshot() -> [DiagnosticEvent] {
        lock.lock(); defer { lock.unlock() }
        return events
    }

    func clear() {
        lock.lock(); events.removeAll(); persistLocked(); lock.unlock()
        NotificationCenter.default.post(name: Self.changed, object: self)
    }

    private func append(_ level: DiagnosticLevel, _ message: String) {
        lock.lock()
        if let previous = events.last, previous.level == level, previous.message == message,
           Date().timeIntervalSince(previous.date) < 10 {
            lock.unlock()
            return
        }
        events.append(DiagnosticEvent(date: Date(), level: level, message: message))
        events = events.filter { $0.date >= Date().addingTimeInterval(-retention) }
        if events.count > maximumEvents { events.removeFirst(events.count - maximumEvents) }
        persistLocked()
        lock.unlock()
        switch level {
        case .info: logger.info("\(message, privacy: .public)")
        case .warning: logger.warning("\(message, privacy: .public)")
        case .error: logger.error("\(message, privacy: .public)")
        }
        NotificationCenter.default.post(name: Self.changed, object: self)
    }
    private func persistLocked() { UserDefaults.standard.set(try? JSONEncoder().encode(events), forKey: storageKey) }
}
